openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -649,114 +605,116 @@ def _get_library_component_tags_count(library_key) -> dict:
+    # version still exists; and (b) we might want to make some queries on the
+    # block even after it's been removed, since there might be versioned
+    # references to it.
+    draft_version = component.versioning.draft
+    if not draft_version:
+        raise ContentLibraryBlockNotFound(usage_key)
+
+    published_version = component.versioning.published
+
     return LibraryXBlockMetadata(
         usage_key=usage_key,
-        def_key=def_key,
-        display_name=get_block_display_name(def_key),
-        has_unpublished_changes=lib_bundle.does_definition_have_unpublished_changes(def_key),
+        display_name=draft_version.title,
+        has_unpublished_changes=(draft_version != published_version),
     )
 
 
-def get_library_block_olx(usage_key):
-    """
-    Get the OLX source of the given XBlock.
-    """
-    assert isinstance(usage_key, LibraryUsageLocatorV2)
-    definition_key = get_library_block(usage_key).def_key
-    xml_str = get_bundle_file_data(
-        bundle_uuid=definition_key.bundle_uuid,  # pylint: disable=no-member
-        path=definition_key.olx_path,  # pylint: disable=no-member
-        use_draft=DRAFT_NAME,
-    ).decode('utf-8')
-    return xml_str
-
-
 def set_library_block_olx(usage_key, new_olx_str):
     """
     Replace the OLX source of the given XBlock.
+
     This is only meant for use by developers or API client applications, as
     very little validation is done and this can easily result in a broken XBlock
     that won't load.
     """
     # because this old pylint can't understand attr.ib() objects, pylint: disable=no-member
     assert isinstance(usage_key, LibraryUsageLocatorV2)
+
     # Make sure the block exists:
-    metadata = get_library_block(usage_key)
-    block_type = usage_key.block_type
+    _block_metadata = get_library_block(usage_key)
+
     # Verify that the OLX parses, at least as generic XML:
     node = etree.fromstring(new_olx_str)
-    if node.tag != block_type:
-        raise ValueError(f"Invalid root tag in OLX, expected {block_type}")
-    # Write the new XML/OLX file into the library bundle's draft
-    draft = get_or_create_bundle_draft(metadata.def_key.bundle_uuid, DRAFT_NAME)
-    write_draft_file(draft.uuid, metadata.def_key.olx_path, new_olx_str.encode('utf-8'))
-    # Clear the bundle cache so everyone sees the new block immediately:
-    BundleCache(metadata.def_key.bundle_uuid, draft_name=DRAFT_NAME).clear()
+
+    # We're intentionally NOT checking if the XBlock type is installed, since
+    # this is one of the only tools you can reach for to edit content for an
+    # XBlock that's broken or missing.
+    component = get_component_from_usage_key(usage_key)
+
+    # Get the title from the new OLX (or default to the default specified on the
+    # XBlock's display_name field.
+    new_title = node.attrib.get(
+        "display_name",
+        xblock_type_display_name(usage_key.block_type),
+    )
```

Comment:
^ Should check that the node tag matches the block type.

Reply:
Done. Now raises a ValueError if it doesn't match.